service/dap/server_test.go

```diff
@@ -4562,41 +4555,41 @@ func TestBadLaunchRequests(t *testing.T) {
 			"Failed to launch: The program attribute is missing in debug configuration.")
 
 		// Bad "args"
-		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "exec", "program": fixture.Path, "args": nil})
+		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "exec", "program": fixture.Path, "args": "foobar"})
 		checkFailedToLaunchWithMessage(client.ExpectInvisibleErrorResponse(t),
-			"Failed to launch: 'args' attribute '<nil>' in debug configuration is not an array.")
+			"Failed to launch: invalid debug configuration - cannot unmarshal string into \"args\" of type []string")
 
 		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "exec", "program": fixture.Path, "args": 12345})
 		checkFailedToLaunchWithMessage(client.ExpectInvisibleErrorResponse(t),
-			"Failed to launch: 'args' attribute '12345' in debug configuration is not an array.")
+			"Failed to launch: invalid debug configuration - cannot unmarshal number into \"args\" of type []string")
 
 		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "exec", "program": fixture.Path, "args": []int{1, 2}})
 		checkFailedToLaunchWithMessage(client.ExpectInvisibleErrorResponse(t),
-			"Failed to launch: value '1' in 'args' attribute in debug configuration is not a string.")
+			"Failed to launch: invalid debug configuration - cannot unmarshal number into \"args\" of type string")
 
 		// Bad "buildFlags"
 		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "buildFlags": 123})
 		checkFailedToLaunchWithMessage(client.ExpectInvisibleErrorResponse(t),
-			"Failed to launch: 'buildFlags' attribute '123' in debug configuration is not a string.")
+			"Failed to launch: invalid debug configuration - cannot unmarshal number into \"buildFlags\" of type string")
 
 		client.LaunchRequestWithArgs(map[string]interface{}{"mode": "debug", "program": fixture.Source, "substitutePath": 123})
 		checkFailedToLaunchWithMessage(client.ExpectInvisibleErrorResponse(t),
-			"Failed to launch: 'substitutePath' attribute '123' in debug configuration is not a []{'from': string, 'to': string}")
+			"Failed to launch: invalid debug configuration - cannot unmarshal number into \"substitutePath\" of type []dap.SubstitutePath")
```

Comment:
Agree. Added a special casing for substitutePath and named the helper to be launch/attach request-specific (now unmarshalLaunchAttachArgs) because substitutePath is only for launch/attach requests.

fyi I am not sure how much more effort we want to make here without understanding how/why people would end up with such misconfiguration (assigning a completely wrong type to substitutePath) in practice. I think, it's better for us to focus on improving the documentation of how to use substitutePath (e.g. what does "" mean? Is it prefix of the path? should it use platform-specific paths? ....), and then as we gather more common misconfiguration cases over time we tweak the error messages to help minimizing confusion.

I think error messages when users specify non-string array value as args (e.g. If user uses [1, 2] instead of ["1", "2"], the error message is `cannot unmarshal number into \"args\" of type string`) is also confusing (encoding/json error message doesn't keep the full context of parsing and there are multiple issues open about improving json error messages) and needs improvement if it turns out many users end up with misconfiguration :-(

Reply:
There was a bug in unmarshalLaunchAttachArgs (should've checked Field instead of Value of json UnmarshalTypeError) and fixed.